lib/jake.ex

```diff
@@ -9,31 +9,90 @@ defmodule Jake do
     "string"
   ]
 
-  def gen(%{"anyOf" => options} = spec) when is_list(options) do
-    Enum.map(options, fn option ->
-      gen(Map.merge(Map.drop(spec, ["anyOf"]), option))
-    end)
+  def generator(jschema) do
+    IO.puts(jschema)
+    jschema |> Poison.decode!() |> gen_init()
+  end
+
+  def gen_init(%{"anyOf" => options} = map) when is_list(options) do
+    nmap = Map.drop(map, ["anyOf"])
+
+    for(n <- options, is_map(n), do: Jake.gen_init(Map.merge(nmap, n)))
     |> StreamData.one_of()
   end
 
-  def gen(%{"type" => type} = spec) when is_binary(type) do
-    module = String.to_existing_atom("Elixir.Jake.#{String.capitalize(type)}")
-    apply(module, :gen, [spec])
+  def gen_init(%{"allOf" => options} = map) when is_list(options) do
+    nmap = Map.drop(map, ["allOf"])
+
+    Enum.reduce(options, %{}, fn x, acc -> Jake.MapUtil.deep_merge(acc, x) end)
+    |> Map.merge(nmap)
+    |> Jake.gen_init()
   end
 
-  def gen(%{"type" => types} = spec) when is_list(types) do
-    Enum.map(types, fn type ->
-      gen(%{spec | "type" => type})
-    end)
-    |> StreamData.one_of()
+  def gen_init(map) do
+    gen_all(map, map["enum"], map["type"])
   end
 
-  # type not present
-  def gen(spec) do
-    Enum.map(@types, fn type ->
-      Map.put(spec, "type", type)
-      |> gen()
-    end)
-    |> StreamData.one_of()
+  def gen_all(map, enum, type) when enum != nil, do: gen_enum(enum, type)
+
+  def gen_all(map, enum, type) when is_list(type) do
+    list = for n <- type, do: %{"type" => n}
+    nmap = Map.drop(map, ["type"])
+    for(n <- list, is_map(n), do: Map.merge(n, nmap) |> Jake.gen_init()) |> StreamData.one_of()
+  end
+
+  def gen_all(map, enum, type) when type in @types, do: gen_type(type, map)
+
+  def gen_all(map, enum, type) when type == nil do
+    Jake.Notype.gen_notype(map, type)
+  end
+
+  def gen_type(type, map) when type == "string" do
+    Jake.String.gen_string(map)
+  end
+
+  def gen_type(type, map) when type in ["integer", "number"] do
+    Jake.Number.gen_number(map, type)
+  end
+
+  def gen_type(type, map) when type == "boolean" do
+    StreamData.boolean()
+  end
+
+  def gen_type(type, map) when type == "null" do
+    StreamData.constant(nil)
+  end
+
+  def gen_type(type, map) when type == "array" do
+    Jake.Array.gen_array(map, type)
+  end
+
+  def gen_type(type, map) when type == "object" do
+    Jake.Object.gen_object(map, type)
+  end
+
+  def gen_enum(list, type) do
+    nlist =
+      case type do
+        x when x == "integer" ->
+          nlist = for n <- list, is_integer(n), do: n
+
+        x when x == "number" ->
+          for n <- list, is_number(n), do: n
+
+        x when x == "string" ->
+          for n <- list, is_binary(n), do: n
+
+        x when x == "array" ->
+          for n <- list, is_list(n), do: n
+
+        x when x == "object" ->
+          for n <- list, is_map(n), do: n
+
+        _ ->
+          list
+      end
+
```

Comment:
why do we have to do all these? can't we just use member_of, assuming the given schema is valid?

Reply:
Consider the case when enum list for **type number** consists of null or string items along with regular numbers, then if we use only **member_of** without filtering first, then we may get wrong result, even if it is possible to get correct result. 

You can also check out the [test integer enum](https://github.com/kanishka-linux/genson/blob/master/test/jake_test.exs#L184)

Reply:
I had assumed enum value would be always correct, I don't understand why would anybody use invalid values.

But let's say we follow the spec and validate the value against enclosing spec, we only seem to handle the `type` right? what if it's a complex case which has other constraints like `maximum` etc? Do you think this could be handled in a simpler way?

Reply:
> I had assumed enum value would be always correct

In fact, I had also assumed the same, but later on looked at [schema doc](https://json-schema.org/understanding-json-schema/reference/generic.html#enumerated-values), which has mentioned that 
``the elements in the enum array should also be valid against the enclosing schema:``, therefore, I decided to filter the enum list depending on the schema type.

>But let's say we follow the spec and validate the value against enclosing spec, we only seem to handle the type right? 

yes

>what if it's a complex case which has other constraints like maximum etc? Do you think this could be handled in a simpler way?

I also thought about it, but it is difficult and hence handled only simple case. I think, validating particular value against schema is out of scope of this project and is a separate project in itself. For this we can use something like ``ExJsonSchema.Validator``, in the beginning, to filter out valid enum values, which then can feed to StreamData.

Reply:
> This library could assume that the schema is correct.

It will certainly simplify things.